backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/extraction/impl/LuceneCollectors.java

```diff
@@ -60,10 +75,12 @@ public void collect(IndexSearcher indexSearcher, Query luceneQuery, int offset,
 		}
 		extractTopDocs( offset, limit );
 
-		if ( !requireFieldDocRescoring ) {
-			return;
+		if ( requireFieldDocRescoring ) {
+			handleRescoring( indexSearcher, luceneQuery );
+		}
+		if ( nestedPathsInSort != null && !nestedPathsInSort.isEmpty() ) {
+			extractTopDocsUsingTheirNested( indexSearcher, luceneQuery, offset, limit );
```

Comment:
I moved `extractTopDocsUsingTheirNested` above.
But without other changes we need to call `extractTopDocs` before we call `extractTopDocsUsingTheirNested`, since we need the `ScoreDoc[] scoreDocs` to f etch the `nestedDocumentMap`.

Reply:
Ok... now that I think about it, I think there's a flaw in this approach.
What about the following scenario:

* The query matches documents A, B, C
* The limit is 2
* The first search returns top documents [`A, B]`
* The second search, taking into account nested documents, would have returned `[C, B]` (`A` excluded because it's third), but since `C` was not in the top docs of the first search, it ends up returning `[B, A]`.

I suspect this is a very real possibility...

The only option to avoid this problem would be to revert the logic:

* Currently you perform the `search()` call once, and then if you discover that there are nested sorts, you try to "fix" the first results by running a second `search()`.
* The alternative would be to detect *before any `search()`* that there are nested sorts. In that case you would run a preliminary `search()` to collect nested documents (and only nested documents) *without a limit*. Then you would run the actual `search()` with a limit.

I'm aware this would be require an awful lot of memory for big indexes, but at least it would work (?). We can create a ticket to try and optimize this, maybe.

Reply:
Actually, I've had a look at the Elasticsearch code, and it seems they retrieve nested documents on the fly when it comes to sorts. See my other comment. I think if you manage to implement a solution that does the same, the flaw I just mentioned will just disappear. Could you have a look, please?

Reply:
Sure